### PR TITLE
GH-96793: Change `FOR_ITER` to not pop the iterator on exhaustion.

### PR DESCRIPTION
**Pull request by markshannon**
Requires an additional `POP_TOP` at the end of the loop. By itself, this would be a pointless change, but as part of #96793 it makes sense.

Pyperformance shows a [1% slowdown](https://gist.github.com/markshannon/85cd6a097d7ce699fdad42330733be65) but I suspect that is a bit noisy. The 11% slowdown of `unpack_sequence` makes no sense.
Given the dispatching overhead is in the 10-15% and `FOR_ITER` represents 1-2% of instructions executed we would expect at worst a 0.3% slowdown.

I will get performance numbers for the whole of #96793 before merging the parts.

* Issue: gh-96793

### CODE REVIEW

**Comment by markshannon** [comment]
With the tweaks to skip the `POP_TOP` in specialized forms the performance results are [in the noise](https://gist.github.com/markshannon/a1c374c3e186315ff4cb691857a4ce48).

File: Python/specialize.c [hunk 1]
```diff
@@ -2193,27 +2193,25 @@ int
...
-    }
-    else if (tp == &PyRangeIter_Type && next_op == STORE_FAST) {
-        _Py_SET_OPCODE(*instr, FOR_ITER_RANGE);
-        goto success;
-    }
-    else {
-        SPECIALIZATION_FAIL(FOR_ITER,
-                            _PySpecialization_ClassifyIterator(iter));
-        goto failure;
+    if (_Py_OPCODE(instr[INLINE_CACHE_ENTRIES_FOR_ITER + oparg + 1]) == POP_TOP) {
```
Comment (iritkatriel) [inline Python/specialize.c]: Why would it not be POP_TOP?

Comment (markshannon) [reply Python/specialize.c]: The compiler did something clever.
I don't know what that might be, but I don't want to rely on the compiler not transforming the code.

**Comment by markshannon** [comment]
I'm considering changing the `POP_TOP` to `END_LOOP` which would avoid that problem and some others related to handling generator return values.

**Review by iritkatriel — APPROVED** [review]

**Review by iritkatriel — APPROVED** [review]
LGTM

**Pull request merged**